File: src/test/java/cz/vhromada/catalog/utils/ShowUtils.java
package cz.vhromada.catalog.utils;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.SoftAssertions.assertSoftly;

import java.util.ArrayList;
import java.util.List;

import javax.persistence.EntityManager;

import cz.vhromada.catalog.domain.Genre;
import cz.vhromada.catalog.entity.Show;
import cz.vhromada.common.utils.CollectionUtils;

/**
 * A class represents utility class for shows.
 *
 * @author Vladimir Hromada
 */
public final class ShowUtils {

    /**
     * ID
     */
    public static final int ID = 1;

    /**
     * Count of shows
     */
    public static final int SHOWS_COUNT = 3;

    /**
     * Position
     */
    public static final int POSITION = 10;

    /**
     * Show name
     */
    private static final String SHOW = "Show ";

    /**
     * Creates a new instance of ShowUtils.
     */
    private ShowUtils() {
    }

    /**
     * Returns show.
     *
     * @param id ID
     * @return show
     */
    public static cz.vhromada.catalog.domain.Show newShowDomain(final Integer id) {
        final cz.vhromada.catalog.domain.Show show = new cz.vhromada.catalog.domain.Show();
        updateShow(show);
        show.setGenres(CollectionUtils.newList(GenreUtils.newGenreDomain(id)));
        show.setPicture(id);
        if (id != null) {
            show.setId(id);
            show.setPosition(id - 1);
        } else {
            show.setPosition(0);
        }
        show.setSeasons(new ArrayList<>());

        return show;
    }

    /**
     * Returns show with seasons.
     *
     * @param id ID
     * @return show with seasons
     */
    public static cz.vhromada.catalog.domain.Show newShowWithSeasons(final Integer id) {
        final cz.vhromada.catalog.domain.Show show = newShowDomain(id);
        show.setSeasons(CollectionUtils.newList(SeasonUtils.newSeasonWithEpisodes(id)));

        return show;
    }

    /**
     * Updates show fields.
     *
     * @param show show
     */
    @SuppressWarnings("Duplicates")
    public static void updateShow(final cz.vhromada.catalog.domain.Show show) {
        show.setCzechName("czName");
        show.setOriginalName("origName");
        show.setCsfd("Csfd");
        show.setImdbCode(1000);
        show.setWikiEn("enWiki");
        show.setWikiCz("czWiki");
        show.setNote("Note");
    }

    /**
     * Returns show.
     *
     * @param id ID
     * @return show
     */
    public static Show newShow(final Integer id) {
        final Show show = new Show();
        updateShow(show);
        show.setGenres(CollectionUtils.newList(GenreUtils.newGenre(id)));
        show.setPicture(id);
        if (id != null) {
            show.setId(id);
            show.setPosition(id - 1);
        } else {
            show.setPosition(0);
        }

        return show;
    }

    /**
     * Updates show fields.
     *
     * @param show show
     */
    @SuppressWarnings("Duplicates")
    public static void updateShow(final Show show) {
        show.setCzechName("czName");
        show.setOriginalName("origName");
        show.setCsfd("Csfd");
        show.setImdbCode(1000);
        show.setWikiEn("enWiki");
        show.setWikiCz("czWiki");
        show.setNote("Note");
    }

    /**
     * Returns shows.
     *
     * @return shows
     */
    public static List<cz.vhromada.catalog.domain.Show> getShows() {
        final List<cz.vhromada.catalog.domain.Show> shows = new ArrayList<>();
        for (int i = 0; i < SHOWS_COUNT; i++) {
            shows.add(getShow(i + 1));
        }

        return shows;
    }

    /**
     * Returns show for index.
     *
     * @param index index
     * @return show for index
     */
    public static cz.vhromada.catalog.domain.Show getShow(final int index) {
        final int imdbMultiplier = 100;

        final cz.vhromada.catalog.domain.Show show = new cz.vhromada.catalog.domain.Show();
        show.setId(index);
        show.setCzechName(SHOW + index + " czech name");
        show.setOriginalName(SHOW + index + " original name");
        show.setCsfd(SHOW + index + " CSFD");
        show.setImdbCode(index * imdbMultiplier);
        show.setWikiEn(SHOW + index + " English Wikipedia");
        show.setWikiCz(SHOW + index + " Czech Wikipedia");
        show.setPicture(MovieUtils.MOVIES_COUNT + index);
        show.setNote(index == 2 ? SHOW + "2 note" : "");
        show.setPosition(index - 1);
        final List<Genre> genres = new ArrayList<>();
        genres.add(GenreUtils.getGenreDomain(index));
        if (index == 3) {
            genres.add(GenreUtils.getGenreDomain(4));
        }
        show.setGenres(genres);
        show.setSeasons(SeasonUtils.getSeasons(index));

        return show;
    }

    /**
     * Returns show.
     *
     * @param entityManager entity manager
     * @param id            show ID
     * @return show
     */
    public static cz.vhromada.catalog.domain.Show getShow(final EntityManager entityManager, final int id) {
        return entityManager.find(cz.vhromada.catalog.domain.Show.class, id);
    }

    /**
     * Returns show with updated fields.
     *
     * @param id            show ID
     * @param entityManager entity manager
     * @return show with updated fields
     */
    public static cz.vhromada.catalog.domain.Show updateShow(final EntityManager entityManager, final int id) {
        final cz.vhromada.catalog.domain.Show show = getShow(entityManager, id);
        updateShow(show);
        show.setGenres(CollectionUtils.newList(GenreUtils.getGenreDomain(1)));
        show.setPosition(POSITION);

        return show;
    }

    /**
     * Returns count of shows.
     *
     * @param entityManager entity manager
     * @return count of shows
     */
    public static int getShowsCount(final EntityManager entityManager) {
        return entityManager.createQuery("SELECT COUNT(s.id) FROM Show s", Long.class).getSingleResult().intValue();
    }

    /**
     * Asserts shows deep equals.
     *
     * @param expected expected shows
     * @param actual   actual shows
     */
    public static void assertShowsDeepEquals(final List<cz.vhromada.catalog.domain.Show> expected, final List<cz.vhromada.catalog.domain.Show> actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertThat(expected.size()).isEqualTo(actual.size());
        if (!expected.isEmpty()) {
            for (int i = 0; i < expected.size(); i++) {
                assertShowDeepEquals(expected.get(i), actual.get(i));
            }
        }
    }

    /**
     * Asserts show deep equals.
     *
     * @param expected expected show
     * @param actual   actual show
     */
    public static void assertShowDeepEquals(final cz.vhromada.catalog.domain.Show expected, final cz.vhromada.catalog.domain.Show actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertSoftly(softly -> {
            softly.assertThat(expected.getId()).isEqualTo(actual.getId());
            softly.assertThat(expected.getCzechName()).isEqualTo(actual.getCzechName());
            softly.assertThat(expected.getOriginalName()).isEqualTo(actual.getOriginalName());
            softly.assertThat(expected.getCsfd()).isEqualTo(actual.getCsfd());
            softly.assertThat(expected.getImdbCode()).isEqualTo(actual.getImdbCode());
            softly.assertThat(expected.getWikiEn()).isEqualTo(actual.getWikiEn());
            softly.assertThat(expected.getWikiCz()).isEqualTo(actual.getWikiCz());
            softly.assertThat(expected.getPicture()).isEqualTo(actual.getPicture());
            softly.assertThat(expected.getNote()).isEqualTo(actual.getNote());
            softly.assertThat(expected.getPosition()).isEqualTo(actual.getPosition());
            GenreUtils.assertGenresDeepEquals(expected.getGenres(), actual.getGenres());
            if (expected.getSeasons() == null) {
                softly.assertThat(actual.getSeasons()).isNull();
            } else {
                SeasonUtils.assertSeasonsDeepEquals(expected.getSeasons(), actual.getSeasons());
            }
        });
    }

    /**
     * Asserts shows deep equals.
     *
     * @param expected expected shows
     * @param actual   actual shows
     */
    public static void assertShowListDeepEquals(final List<Show> expected, final List<cz.vhromada.catalog.domain.Show> actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertThat(expected.size()).isEqualTo(actual.size());
        if (!expected.isEmpty()) {
            for (int i = 0; i < expected.size(); i++) {
                assertShowDeepEquals(expected.get(i), actual.get(i));
            }
        }
    }

    /**
     * Asserts show deep equals.
     *
     * @param expected expected show
     * @param actual   actual show
     */
    public static void assertShowDeepEquals(final Show expected, final cz.vhromada.catalog.domain.Show actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertSoftly(softly -> {
            softly.assertThat(actual.getId()).isEqualTo(expected.getId());
            softly.assertThat(actual.getCzechName()).isEqualTo(expected.getCzechName());
            softly.assertThat(actual.getOriginalName()).isEqualTo(expected.getOriginalName());
            softly.assertThat(actual.getCsfd()).isEqualTo(expected.getCsfd());
            softly.assertThat(actual.getImdbCode()).isEqualTo(expected.getImdbCode());
            softly.assertThat(actual.getWikiEn()).isEqualTo(expected.getWikiEn());
            softly.assertThat(actual.getWikiCz()).isEqualTo(expected.getWikiCz());
            softly.assertThat(actual.getPicture()).isEqualTo(expected.getPicture());
            softly.assertThat(actual.getNote()).isEqualTo(expected.getNote());
            softly.assertThat(actual.getPosition()).isEqualTo(expected.getPosition());
            GenreUtils.assertGenreListDeepEquals(expected.getGenres(), actual.getGenres());
        });
    }

}

File: src/test/kotlin/cz/vhromada/catalog/utils/ShowUtils.kt
package cz.vhromada.catalog.utils

import cz.vhromada.catalog.domain.Genre
import cz.vhromada.catalog.entity.Show
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.SoftAssertions.assertSoftly
import javax.persistence.EntityManager

/**
 * Updates show fields.
 *
 * @return updated show
 */
fun cz.vhromada.catalog.domain.Show.updated(): cz.vhromada.catalog.domain.Show {
    return copy(czechName = "czName", originalName = "origName", csfd = "Csfd", imdbCode = 1000, wikiEn = "enWiki", wikiCz = "czWiki", note = "Note")
}

/**
 * Updates show fields.
 *
 * @return updated show
 */
fun Show.updated(): Show {
    return copy(czechName = "czName", originalName = "origName", csfd = "Csfd", imdbCode = 1000, wikiEn = "enWiki", wikiCz = "czWiki", note = "Note")
}

/**
 * A class represents utility class for shows.
 *
 * @author Vladimir Hromada
 */
object ShowUtils {

    /**
     * Count of shows
     */
    const val SHOWS_COUNT = 3

    /**
     * Position
     */
    const val POSITION = 10

    /**
     * Show name
     */
    private const val SHOW = "Show "

    /**
     * Returns shows.
     *
     * @return shows
     */
    fun getShows(): List<cz.vhromada.catalog.domain.Show> {
        val shows = mutableListOf<cz.vhromada.catalog.domain.Show>()
        for (i in 0 until SHOWS_COUNT) {
            shows.add(getShow(i + 1))
        }

        return shows
    }

    /**
     * Returns show.
     *
     * @param id ID
     * @return show
     */
    fun newShowDomain(id: Int?): cz.vhromada.catalog.domain.Show {
        return cz.vhromada.catalog.domain.Show(
                id = id,
                czechName = "",
                originalName = "",
                csfd = null,
                imdbCode = null,
                wikiEn = null,
                wikiCz = null,
                picture = id,
                note = null,
                position = if (id == null) null else id - 1,
                genres = listOf(GenreUtils.newGenreDomain(id)),
                seasons = emptyList())
                .updated()
    }

    /**
     * Returns show with seasons.
     *
     * @param id ID
     * @return show with seasons
     */
    fun newShowWithSeasons(id: Int?): cz.vhromada.catalog.domain.Show {
        return newShowDomain(id)
                .copy(seasons = listOf(SeasonUtils.newSeasonWithEpisodes(id)))
    }

    /**
     * Returns show.
     *
     * @param id ID
     * @return show
     */
    fun newShow(id: Int?): Show {
        return Show(
                id = id,
                czechName = "",
                originalName = "",
                csfd = null,
                imdbCode = null,
                wikiEn = null,
                wikiCz = null,
                picture = id,
                note = null,
                position = if (id == null) null else id - 1,
                genres = listOf(GenreUtils.newGenre(id)))
                .updated()
    }

    /**
     * Returns show for index.
     *
     * @param index index
     * @return show for index
     */
    fun getShow(index: Int): cz.vhromada.catalog.domain.Show {
        val imdbMultiplier = 100
        val genres = mutableListOf<Genre>()
        genres.add(GenreUtils.getGenreDomain(index))
        if (index == 3) {
            genres.add(GenreUtils.getGenreDomain(4))
        }

        return cz.vhromada.catalog.domain.Show(
                id = index,
                czechName = "$SHOW$index czech name",
                originalName = "$SHOW$index original name",
                csfd = "$SHOW$index CSFD",
                imdbCode = index * imdbMultiplier,
                wikiEn = "$SHOW$index English Wikipedia",
                wikiCz = "$SHOW$index Czech Wikipedia",
                picture = MovieUtils.MOVIES_COUNT + index,
                note = if (index == 2) SHOW + "2 note" else "",
                position = index - 1,
                genres = genres,
                seasons = SeasonUtils.getSeasons(index))
    }

    /**
     * Returns show.
     *
     * @param entityManager entity manager
     * @param id            show ID
     * @return show
     */
    fun getShow(entityManager: EntityManager, id: Int): cz.vhromada.catalog.domain.Show? {
        return entityManager.find(cz.vhromada.catalog.domain.Show::class.java, id)
    }

    /**
     * Returns show with updated fields.
     *
     * @param id            show ID
     * @param entityManager entity manager
     * @return show with updated fields
     */
    fun updateShow(entityManager: EntityManager, id: Int): cz.vhromada.catalog.domain.Show {
        return getShow(entityManager, id)!!
                .updated()
                .copy(position = POSITION, genres = mutableListOf(GenreUtils.getGenreDomain(1)))
    }

    /**
     * Returns count of shows.
     *
     * @param entityManager entity manager
     * @return count of shows
     */
    @Suppress("CheckStyle")
    fun getShowsCount(entityManager: EntityManager): Int {
        return entityManager.createQuery("SELECT COUNT(s.id) FROM Show s", java.lang.Long::class.java).singleResult.toInt()
    }

    /**
     * Asserts shows deep equals.
     *
     * @param expected expected shows
     * @param actual   actual shows
     */
    fun assertShowsDeepEquals(expected: List<cz.vhromada.catalog.domain.Show?>?, actual: List<cz.vhromada.catalog.domain.Show?>?) {
        assertSoftly {
            it.assertThat(expected).isNotNull
            it.assertThat(actual).isNotNull
        }
        assertThat(expected!!.size).isEqualTo(actual!!.size)
        if (expected.isNotEmpty()) {
            for (i in expected.indices) {
                assertShowDeepEquals(expected[i], actual[i])
            }
        }
    }

    /**
     * Asserts show deep equals.
     *
     * @param expected expected show
     * @param actual   actual show
     */
    fun assertShowDeepEquals(expected: cz.vhromada.catalog.domain.Show?, actual: cz.vhromada.catalog.domain.Show?) {
        assertSoftly {
            it.assertThat(expected).isNotNull
            it.assertThat(actual).isNotNull
        }
        assertSoftly {
            it.assertThat(expected!!.id).isEqualTo(actual!!.id)
            it.assertThat(expected.czechName).isEqualTo(actual.czechName)
            it.assertThat(expected.originalName).isEqualTo(actual.originalName)
            it.assertThat(expected.csfd).isEqualTo(actual.csfd)
            it.assertThat(expected.imdbCode).isEqualTo(actual.imdbCode)
            it.assertThat(expected.wikiEn).isEqualTo(actual.wikiEn)
            it.assertThat(expected.wikiCz).isEqualTo(actual.wikiCz)
            it.assertThat(expected.picture).isEqualTo(actual.picture)
            it.assertThat(expected.note).isEqualTo(actual.note)
            it.assertThat(expected.position).isEqualTo(actual.position)
            GenreUtils.assertGenresDeepEquals(expected.genres, actual.genres)
            SeasonUtils.assertSeasonsDeepEquals(expected.seasons, actual.seasons)
        }
    }

    /**
     * Asserts shows deep equals.
     *
     * @param expected expected shows
     * @param actual   actual shows
     */
    fun assertShowListDeepEquals(expected: List<Show?>?, actual: List<cz.vhromada.catalog.domain.Show?>?) {
        assertSoftly {
            it.assertThat(expected).isNotNull
            it.assertThat(actual).isNotNull
        }
        assertThat(expected!!.size).isEqualTo(actual!!.size)
        if (expected.isNotEmpty()) {
            for (i in expected.indices) {
                assertShowDeepEquals(expected[i], actual[i])
            }
        }
    }

    /**
     * Asserts show deep equals.
     *
     * @param expected expected show
     * @param actual   actual show
     */
    fun assertShowDeepEquals(expected: Show?, actual: cz.vhromada.catalog.domain.Show?) {
        assertSoftly {
            it.assertThat(expected).isNotNull
            it.assertThat(actual).isNotNull
        }
        assertSoftly {
            it.assertThat(actual!!.id).isEqualTo(expected!!.id)
            it.assertThat(actual.czechName).isEqualTo(expected.czechName)
            it.assertThat(actual.originalName).isEqualTo(expected.originalName)
            it.assertThat(actual.csfd).isEqualTo(expected.csfd)
            it.assertThat(actual.imdbCode).isEqualTo(expected.imdbCode)
            it.assertThat(actual.wikiEn).isEqualTo(expected.wikiEn)
            it.assertThat(actual.wikiCz).isEqualTo(expected.wikiCz)
            it.assertThat(actual.picture).isEqualTo(expected.picture)
            it.assertThat(actual.note).isEqualTo(expected.note)
            it.assertThat(actual.position).isEqualTo(expected.position)
            GenreUtils.assertGenreListDeepEquals(expected.genres, actual.genres)
        }
    }

}
